### PR TITLE
Optimize matrix scanning

### PR DESCRIPTION
**Pull request by eltang**
@jackhumbert I've only tested if this commit compiles. Should I flash it to a keyboard and test it out?

### CODE REVIEW

**Comment by jackhumbert** [comment]
Yeah, always test on boards - I don't really wanna merge stuff that I haven't written or thoroughly tested.

The `PIN()` math is cool, but I'd like to keep the notation how it is for now - it's a lot easier to read/change.

I'm interested in what you did to the matrix file though - would you mind explaining what all you're doing? Is it faster/more efficient?

**Comment by eltang** [comment]
That macro is necessary—otherwise, users will have to prefix their pin numbers with `0x`. The code is more efficient now.

I'd be glad. First, notice that conditional jumps have been eliminated. That alone speeds things up. Also, instead of having variable for each pin letter and matching up the variables with the correct registers, the pin letter is instead treated as a hex value and the address of the register is calculated from it. Does that make sense? I'll guess I'll test this code in my keyboard now.

**Comment by eltang** [comment]
Tested and working!

**Comment by eltang** [comment]
This is also why I asked if you would be supporting more chips—the registers might be different.

**Comment by jackhumbert** [comment]
Yeah, if you wanna do the shortcuts to keep the notation the same, that'd be great!

There is a plan for more chips, but we can address (haha) those when we get to them with something like `matrix_atmega32u4.c` that gets chosen by the makefile depending on the chip.

Just a general FYI, I'm gonna hold off on merging stuff like this until I've gotten the firmware finalised for the latest Massdrop - it's a little time sensitive that I get things finished.

**Comment by eltang** [comment]
No efficient firmware for the drop participants? :P

**Comment by jackhumbert** [comment]
To be fair, I don't think execution time is really an issue here :) but yeah, I just don't have time to do the testing required!

**Comment by eltang** [comment]
I moved most of the calculations to compile time. Efficiency!

**Comment by eltang** [comment]
I forgot to mention that this saves 20 bytes of memory.

File: quantum/matrix.c [hunk 1]
```diff
@@ -194,114 +194,61 @@ uint8_t matrix_key_count(void)
...
 #if DIODE_DIRECTION == COL2ROW
-    for(int x = 0; x < MATRIX_COLS; x++) {
-        int col = COLS[x];
+    for (uint16_t col = 0; col < MATRIX_COLS; col++) {
+        const uint8_t col_pin = COL_PINS[col];
 #else
-    for(int x = 0; x < MATRIX_ROWS; x++) {
-        int col = ROWS[x];
+    for (uint16_t col = 0; col < MATRIX_ROWS; col++) {
+        const uint8_t col_pin = ROW_PINS[col];
```
Comment (jackhumbert) [inline quantum/matrix.c]: We might as well call this `output_pin` to accommodate both diode directions.

Comment (eltang) [inline quantum/matrix.c]: Wouldn't that make more conditional compilation directives needed?

Comment (jackhumbert) [inline quantum/matrix.c]: We already have the necessary directives - here and down in `select_row`.

Comment (eltang) [inline quantum/matrix.c]: But those variables are used in code that's currently unconditionally compiled.

Comment (jackhumbert) [inline quantum/matrix.c]: `col_pin` references would become `input_pin`, and `row_pin` references would become `output_pin` - the function names could change to reflect that as well. It'd also make it a little clearer how things are actually working.

Comment (eltang) [inline quantum/matrix.c]: Ah. I feel stupid now.

**Comment by eltang** [comment]
Done.

File: quantum/matrix.c (hunk 1 above)
Comment (eltang) [inline quantum/matrix.c]: Done. I think we should keep `COL2ROW`and `ROW2COL` as they are—people not familiar not familiar with how the matrix works might have trouble figuring out which of those to define.

**Comment by eltang** [comment]
@ezuk I just realized that the changes to `matrix.h` might (read: will) break your custom `matrix.c` file. How should we fix this? I'm not sure I understand the I2C stuff enough to make the changes myself.

**Comment by jackhumbert** [comment]
I think updating the custom matrix.c file should be easy enough - we can worry about that when we revisit this later. Duplicating the matrix.h file is just asking for trouble :)

`COL2ROW` and `ROW2COL` aren't any easier to understand that `COL_OUT` or the like - most people hand-wiring don't understand the way the current flows through the matrix anyway.

**Comment by eltang** [comment]
Hmm. Are you sure? Most of them are able to discern which way the diodes go (I hope) …

**Comment by jackhumbert** [comment]
Diode direction is still tricky for a lot of people, but placing them correctly (usually by referencing a photo) doesn't indicate understanding :) I'll talk to @erez about naming these something different.

**Comment by eltang** [comment]
Ahh! The user you mentioned is someone else!

**Comment by eltang** [comment]
I realized that all I have to change are the function names. I'm doing it.

**Comment by eltang** [comment]
Done.

**Comment by eltang** [comment]
@ezuk The ErgoDox EZ uses column-based scanning, correct? In that case, I may have screwed things up.

**Comment by ezuk** [comment]
It's PRs like this that worry me. :) We're not merging this in?

**Comment by eltang** [comment]
I've got to fix some things. Apparently the Ergodox EZ's diodes are in the flipped orientation, but it's `matrix.c` was changed as little as possible and none of the functions were renamed. I've got to clean things up before reopening this. Please confirm if my suspicion is correct.

**Comment by ezuk** [comment]
@eltang we follow the ErgoDox design, basically. Didn't change anything in terms of diode orientation, which is why qmk works on both the DIY and the EZ versions.

**Comment by eltang** [comment]
So the original ErgoDox has flipped diodes?

**Comment by ezuk** [comment]
This is our canonical schematic: https://github.com/ErgoDox-EZ/docs/blob/master/ErgoDox%20EZ%20Schematic.pdf

**Comment by eltang** [comment]
That settles it then. I'll have to fix up that file.

@jackhumbert In the matrix files, I'm having trouble deciding what to do. Should I a) try to integrate the code for the two diode directions as much as possible and use two words `input` and `output` as much as possible, or should I b) split the code apart as much as possible and use the words `row` and `col` as much as possible?

**Comment by eltang** [comment]
That settles it then. I'll have to fix up that file.

@jackhumbert In the matrix files, I'm having trouble deciding what to do. Should I a) try to integrate the code for the two diode directions as much as possible and use two words `input` and `output` as much as possible, or should I b) split the code apart as much as possible and use the words `row` and `col` as much as possible?

**Comment by eltang** [comment]
@jackhumbert I know that you might be too busy to spend time on this right now, so I'm just leaving this here for the future. In the `unselect_rows` function in the current `matrix.c` file, why are the rows being set to internal pull-up rather than than hi-Z?

**Pull request closed without merging**